### DISCUSSION TITLE
How to use BrowserOnly with ES-only modules

### DISCUSSION THREAD

**Discussion by slorber** [facebook/docusaurus#9435] [Q&A]
Hi @NickGerleman

Moving your question here since it's a better place rather than in the original issue you posted it to (https://github.com/facebook/docusaurus/issues/6520#issuecomment-1771762153)

> Something I ran into when converting a project to ES Modules was that BrowserOnly and importing code for the browser, but not the server, no longer works as documented.
> 
> In CommonJS, browser specific code can be dynamic required synchronously inline with the component provider. But this is async with ESModules, so the existing example is no longer correct.
> 
> I ended up wrapping the component in React.lazy, and adding suspense, which works, but is a little unergonomic, and has different semantics (since the page now renders before the browser only code is available).

---

Follow up in https://github.com/facebook/docusaurus/issues/6520#issuecomment-1773457906

> @slorber this is for an effort to port [yogalayout.com](https://yogalayout.com/) (currently using an ancient version of Gatsby) to Docusaurus. Right now I have it targeting 2.4.
> 
> The website has an interactive component, using an asm.js version of Yoga. In modern Yoga, the JavaScript bindings are WebAssembly instead. The current iteration of WebAssembly best practices interoping with bundlers requires top-level await support, which forces the component to be an ES Module.
> 
> Wholesale porting the existing component led to a non-descriptive error message about server compilation failing. Originally I knew this was due to incompatibility with the WASM builds, but I think now it might be a different incompatibility I haven't been able to suss out (I suspect one of the very old dependencies being pulled in). In the meantime, I have been running this playground only on the client.
> 
> Right now, the [documented way](https://docusaurus.io/docs/docusaurus-core#browseronly) to load a module only in the browser looks like this:
> 
> ```tsx
>     <BrowserOnly fallback={<div>Loading...</div>}>
>       {() => {
>         const LibComponent = require('some-lib');
>         return <LibComponent {...props} />;
>       }}
>     </BrowserOnly>
> ```

> If the component is an ESModule, we cannot synchronously require it, but we must return a component synchronously in the API.
> 
> ```tsx
>     // We can't use this pattern
>     <BrowserOnly fallback={<div>Loading...</div>}>
>       {() => {
>         const LibComponent = await import('some-lib');
>         return <LibComponent {...props} />;
>       }}
>     </BrowserOnly>
> ```
> 
> We cannot use a normal import, or top level await, outside the `BrowserOnly`, since it would then be imported on the server as well.
> 
> We can wrap the import in `React.lazy` to create a sync looking React component that suspends until the import is complete, but this will happen after the module has already loaded, so we need to explicitly handle async state.

**Comment by slorber** [comment]
To answer @@NickGerleman, I'm not sure to understand exactly your problem, and it would help if you created a repro with docusaurus.new, using a real ESM package so that I could tweak it and show you a live example using your ESM lib.

`<BrowserOnly>` is only a convenient helper, nothing super fancy, it just delays the rendering to a `useEffect`, you can also do that in userland and consider your dynamic import similarly to any async React call.

```js
function Component() {
  const [yoga, setYoga] = useState(null);

  useEffect(() => {
    import("yoga").then(setYoga);
  },[]) 
  
  if (!yoga) {
    return false;
  }
  
  return <div>{yoga.xyz()}</div>
}
```

We could probably make it easier to import ES modules by allowing the browser only function to be async, but that would only be a shortcut to the code above.

```js
<BrowserOnly async fallback={<div>Loading...</div>}>
  {async () => {
    const yoga = await import('yoga');
    return <div>{yoga.xyz()}</div>;
  }}
</BrowserOnly>
```

**Comment by NickGerleman** [comment]
I didn’t realize the implementation was this simple. I thought their needed to be some magic to not bundle the imports for server code, but if we are just relying on node I guess dynamic imports at runtime are never parsed by server code.

Semantics with the default webpack configuration are potentially different using dynamic import instead of require. The imported component is potentially split into a separate network loaded bundle, rendered later than the first client render. This means faster initial site load, but dynamic content takes longer to appear.

We already need a server fallback, but it might be expected to show more delayed, which could need different transition UI. I think adapting the API to allow async functions makes a lot of sense, and the implications of split bundle could be documented alongside.

Here’s the stacked commit I am working on, where the Docusaurus index.tsx is mostly stock. Though the example is difficult to run because there is local toolchain needed to build the workspace version of yoga itself.   https://github.com/facebook/yoga/pull/1435

**Reply by slorber** [reply]
> I thought their needed to be some magic

No magic at all 😄 

> Semantics with the default webpack configuration are potentially different using dynamic import instead of require. 

That's true, but we also provide a `configureWebpack` lifecycle that permits you to eventually change the default Docusaurus semantics if you don't like that idea. You'll probably need a Webpack PhD though 😅 it's been a while I haven't touched this so I can't really help.

> I think adapting the API to allow async functions makes a lot of sense, and the implications of split bundle could be documented alongside.

The usage of dynamic imports and different semantics is not really the same topic as `<BrowserOnly>`. This component is really just a tiny convenience React helper, that doesn't even care about ESM / CJS, and that you could have easily written in userland.

CF the abstraction Josh Comeau is proposing in his famous blog post, it's exactly the same:
https://www.joshwcomeau.com/react/the-perils-of-rehydration/

<img width="808" alt="CleanShot 2023-10-23 at 12 32 19@2x" src="https://github.com/facebook/docusaurus/assets/749374/8ea6a74a-e14d-45b0-83e5-17dca195b5ea">


I'm not sure it is a good idea to document all this, even more alongside the `<BrowserOnly>`. The way bundles are loaded is an implementation detail, that may change some day if we want to optimize or adopt a new bundler, and the required usage of some scenarios to import ESM code requires you to have broader knowledge of the CJS/ESM interop.

I'm not even sure it will code split your code to a new bundle, nor that it will delay the loading of that bundle. It's not because there are extra JS files being emitted that they won't be loaded ahead of time before hydration for example. This is not super fresh in my head so can't be 100% sure of any of this.

If your code didn't crash when imported on server then you'd probably don't even need any of this, and you could just import it statically and simply call `yoga.init()` in a `useEffect`. Maybe you could create a little wrapper CJS interface to your Wasm code to make the integration easier.

---

In case that helps, I know Relay integrated a Wasm playground in their website. It didn't require `<BrowserOnly>`
https://github.com/facebook/relay/blob/main/website/src/pages/compiler-explorer.js

**Reply by NickGerleman** [reply]
> If your code didn't crash when imported on server then you'd probably don't even need any of this

Yes, this is the root issue here. At first, I couldn’t do SSR, because I was using per-platform export maps, and IIRC the SSR build was unhappy with Node specific APIs being imported.

Now, the same code is being used for both web and node, but we see a “minified react error” without a stack that isn’t very useful. I tried using some docusaurus build flag to avoid minification which didn’t seem to help.

I suspect this is related to the dependencies being brought in by the existing client only playground being ported over instead of the wasm though. Need to check that when I’m back from PTO.

**Reply by NickGerleman** [reply]
> In case that helps, I know Relay integrated a Wasm playground in their website. It didn't require <BrowserOnly>
https://github.com/facebook/relay/blob/main/website/src/pages/compiler-explorer.js

This part of that file looks suspect 😅

```jsx
const Editor = useMemo(() => {
    // Loading the Editor component causes Docusaurus' build time pre-rendering to
    // crash, so we initializie it lazily.
    return require('../compiler-explorer/Editor').default;
  }, []);
```

But it looks like wasm loading is client only: https://github.com/facebook/relay/blob/059b303199fe514c2323da77cf13a35f6b0cd988/website/src/pages/compiler-explorer.js#L268

I think this case is being used for async compilation of the wasm though, where Yoga is instead using top-level await for so we can expose the API synchronously.

**Reply by slorber** [reply]
> Now, the same code is being used for both web and node, but we see a “minified react error” without a stack that isn’t very useful. I tried using some docusaurus build flag to avoid minification which didn’t seem to help.

If you run `docusaurus build --dev` it's supposed to use React dev build and show unminified React errors. But it's a new Docusaurus v3 option (now in RC).

> This part of that file looks suspect 😅

Yes, not sure what you are trying to do here 😅 

It's hard to help you without seeing full but minimal code (the pr you linked too is a bit hard to understand for me). 

If you can show me how to import and use your wasm code to render a hello world or something, I could try to show you how to integrate it in docusaurus.new. Now it's a bit complicated because I only see partial information of many attempts 🤪

**Reply by NickGerleman** [reply]
> If you run docusaurus build --dev it's supposed to use React dev build and show unminified React errors. But it's a new Docusaurus v3 option (now in RC).

Ooh I think I might try to do this, and move to V3 so I can see the exact error during SSR. The site is being newly built anyway.

Otherwise I do have the playground working client only. But I added a fade animation, since experimentally, the component makes it into a different bundle, and we can see a delay with it loading over a simulated slow connection (even after hydration).

**Reply by slorber** [reply]
> Otherwise I do have the playground working client only. 

It should also work at build time 😄 


> But I added a fade animation, since experimentally, the component makes it into a different bundle, and we can see a delay with it loading over a simulated slow connection (even after hydration).

The fact that it is in a different bundle does not necessarily imply the loading will be delayed.

If we have one big `main.js` or 2 files `main.js` + `compiler-explorer.js`, as far as I remember we probably load all of those files eagerly before even starting to hydrate React. Unless you split bundles in tons of small files, I doubt it has a significant perf impact. Static ES imports might also be code split.

If your lib can't support SSR, then you can't avoid having to display a fallback. React first client-side render must match the JSX/HTML of the SSR output exactly, so it's not possible to hack this and attempt to make your editor load faster, give a better UX, and produce less jank, apart from adding proper SSR support for it.

See for example on Relay, the wasm compiler is not loaded instantly: https://relay.dev/compiler-explorer/

Similarly, Mermaid diagrams do not support SSR yet, and we can't avoid the many annoying layout shifts on this page: https://docusaurus.io/tests/pages/diagrams

This UX is not really related to the code splitting, but rather the lack of proper SSR support
